Automattic/newspack-plugin
# feat(subscriptions): add cancellation reason metadata

### All Submissions:






### Changes proposed in this Pull Request:



Closes https://app.asana.com/0/1208702687864922/1208702687865961

This PR adds meta data to cancelled woo subscriptions and uses this new meta to sync cancellation reason to ESPs for cancelled subscriptions.

### How to test the changes in this Pull Request:

Before testing be sure to set the feature flag in wp-config: `define( 'NEWSPACK_SUBSCRIPTIONS_EXPIRATION', true );`. Also ensure Newspack logging is active on the site: `define( 'WP_NEWSPACK_DEBUG', true )`

1. As admin ensure an ESP is connected and sync is working
2. Go to Newspack > Engagement > Show Advanced Settings > Sync contacts to ESP and ensure `Subscription Cancellation Reason` is checked:
![Screenshot 2024-11-22 at 16 21 40](https://github.com/user-attachments/assets/70db908a-f2d7-45f2-bdc7-5b872caeed87)
3. As a reader purchase a new subscription. **This should be the only active subscription for the reader since our ESP sync will only sync cancelled subscription data if there are no other active subscriptions**
4. As admin, go to the edit subscription page for the above subscription and cancel (WooCommerce > Subscriptions)
5. After page reload scroll down to the custom fields section of the edit subscription page and verify a new cancellation reason field exists with a value of `manually-cancelled`:
![Screenshot 2024-11-22 at 14 36 22](https://github.com/user-attachments/assets/d9f2d323-543d-430c-8b3a-68e456597323)
6. As a reader again, purchase another subscription
7. Still as reader go to my account and cancel the subscription (My Account > Subscriptions > View then click the cancel button)
8. Via wp cli, update the pending cancellation subscription end date to one minute from now:

```
$subscription = wcs_get_subscription( SUBSCRIPTION_ID );
$subscription->update_dates( array( 'end_date' => "2024-11-22 19:35:01" ) );
```
9. Wait for the end date time to lapse, then visit the site as the reader to ensure the transition from pending cancel to cancel happens
10. As admin, visit the edit subscriptions page for the subscription and verify the cancellation reason meta field exists with a value of `user-cancelled`
![Screenshot 2024-11-22 at 14 35 35](https://github.com/user-attachments/assets/aa7eb4f8-551f-41b8-a83c-3e8cc4d42571)
 11. Check the logs and ensure the `Subscription Cancellation Reason` field with the expected value is present in the normalized ESP sync data:

![Screenshot 2024-12-02 at 19 00 37](https://github.com/user-attachments/assets/dab4b514-1b68-4ea5-b9dc-1ec920a329a6)


### Other information:
